src/components/LibrariesPage.tsx

```diff
@@ -76,14 +71,64 @@ export class LibrariesPage extends React.Component<LibrariesPageProps, Libraries
+    // There might be QA libraries available, but the user only wants to see the production ones:
+    let filteredLibraries = !this.state.qa && allLibraries && this.filter(allLibraries);
+    // The user has submitted a search, and there are results:
+    let allResults = !this.state.showAll && this.props.results && this.props.results.libraries;
+    // The user has submitted a search, but only wants to see results for production libraries:
+    let filteredResults = !this.state.qa && allResults && this.filter(allResults);
+    // The user has submitted a search, but there are no results:
+    let noResults = !this.state.showAll && !this.props.results && [];
+
+    // There are six possible scenarios:
+    // 1) A search has been submitted, QA mode is active, and there are search results. Set the libraries variable to the list of results.
+    // 2) A search has been submitted, QA mode is not active, and there are search results.  Set the libraries variable to a filtered list
+    // of results--just the ones that are in production.
+    // 3) A search has been submitted, and there are not search results. Set the libraries variable to an empty array;
+    // this will generate a "no results were found" message.
+    // 4) A search has not been submitted, the registry has libraries, and QA mode is not active.  Set the libraries
+    // variable to a filtered list of libraries--just the ones that are in production.
+    // 5) A search has not been submitted, the registry has libraries, and QA mode is active. Set the libraries variable
+    // to the full list of the registry's libraries.
+    // 6) A search has not been submitted, and the registry does not have libraries. Let the libraries
+    // variable remain undefined; this will generate a "no libraries in this registry" message.
+    return (filteredResults || allResults || noResults) || filteredLibraries || allLibraries;
+  }
+
+  async toggleQA(showQA: boolean) {
+    this.setState({ "qa": showQA });
+    let hasAlreadyLoadedQA: boolean = this.props.libraries.libraries && !(this.props.libraries.libraries.every(l => l.stages.registry_stage === "production"));
+
+    if (!hasAlreadyLoadedQA) {
+      let fetch = showQA ? this.props.fetchQA : this.props.fetchData;
+      await fetch();
+    }
+    if (showQA && this.state.searchTerm && !this.props.results) {
+      // The user searched for a QA library, but didn't switch to QA mode until after submitting the search.
+      let searchTerm = new (window as any).FormData;
+      searchTerm.append("name", this.state.searchTerm);
+      await this.search(searchTerm);
+    }
+  }
+
   async search(data: FormData) {
     this.setState({ showAll: false, searchTerm: (data.get("name") as string) });
     await this.props.search(data);
   }
 
   async clear() {
     this.setState({ showAll: true, searchTerm: "" });
-    await this.props.fetchData();
+    let fetch = this.state.qa ? this.props.fetchQA : this.props.fetchData;
+    await fetch();
```

Comment:
So when you search for a term, it's making a new API call and clearing the store of the libraries. That's why there's a need to refetch the QA or production libraries endpoint again? I forgot that this made an API call when clearing the search, I thought it was just un-filtering the data set to show all the libraries again.

Reply:
Yes.  Do you think it would be better to just have the search happen on the front end (i.e. be a filter)?  I did it the way it is because there was already a search method in the server code, so it seemed clunky to duplicate the existing functionality on the front end.  But on the other hand, moving the search to the front end would make some things much easier.

Reply:
Given the API endpoint that does the search for you and not wanting to duplicate functionality, I think it's okay. Everything on the front-end has its pros and cons and for this, relying on the server code is fine. You'd have multiple calls to search and clear (and then possibly switch lists):
![Screen Shot 2019-06-04 at 11 40 32 AM](https://user-images.githubusercontent.com/1280564/58893166-951a9900-86bd-11e9-9b13-f00e5fab7717.png)

It works fine now and I would say leave this part as is. This is not really part of the PR anyway, but if you want to rethink the architecture for this later that'd be a different convo.